File: src/medium/FirstMissingPositive.java
package src.medium;

import java.util.*;

/**
 * https://leetcode.com/problems/first-missing-positive/
 */
public class FirstMissingPositive {

    public static void main(String[] args) {

        int[] nums = {1,2,3,4,5,6,7,8,9,10,11,12,23,20};
        System.out.println(firstMissingPositive(nums));
    }

    public static int firstMissingPositive(int[] nums) {

        TreeSet<Integer> treeSet = new TreeSet<>();
        for (int n : nums) if (n > 0) treeSet.add(n);
        if (treeSet.isEmpty() || treeSet.first() > 1) return 1;
        Iterator<Integer> it = treeSet.iterator();
        int x = it.next();
        int y;
        while(it.hasNext()){
            y = it.next();
            if(y - x >= 2) return x + 1;
            x = y;
        }

        return treeSet.last() + 1;
    }

}

File: src/scala/FirstMissingPositive.scala
package src.scala

/**
 * @author Prashant Sharma
 *         https://leetcode.com/problems/first-missing-positive/
 */
object FirstMissingPositive {

  def main(args: Array[String]): Unit = {
    val nums = Array(1, 2, 0)
    println(firstMissingPositive(nums))
  }

  def firstMissingPositive(nums: Array[Int]): Int = {
    val list = nums.filter(i => i > 0).sorted
    if (list.isEmpty || list(0) > 1) return 1
    for (i <- 0 until list.length - 1) if (list(i + 1) - list(i) >= 2) return list(i) + 1
    list(list.length - 1) + 1
  }
}
